src/command.rs

```diff
@@ -424,34 +425,533 @@ impl <U:WebDriverExtensionRoute> ToJson for WebDriverMessage<U> {
+                }
+            }
+            _ => return Err(WebDriverError::new(ErrorStatus::InvalidArgument,
+                                                "pageLoadStrategy was not a string"))
+        }
+        Ok(())
+    }
+
+    fn validate_proxy(&self, proxy_value: &Json) -> WebDriverResult<()> {
+        let obj = try_opt!(proxy_value.as_object(),
+                           ErrorStatus::InvalidArgument,
+                           "proxy was not an object");
+        for (key, value) in obj.iter() {
+            match &**key {
+                "proxyType" => match value.as_string() {
+                    Some("pac") |
+                    Some("noproxy") |
+                    Some("autodetect") |
+                    Some("system") |
+                    Some("manual") => {},
+                    Some(x) => return Err(WebDriverError::new(
+                        ErrorStatus::InvalidArgument,
+                        format!("{} was not a valid proxyType value", x))),
+                    None => return Err(WebDriverError::new(
+                        ErrorStatus::InvalidArgument,
+                        "proxyType value was not a string")),
+                },
+                "proxyAutoconfigUrl" => match value.as_string() {
+                    Some(x) => {
+                        try!(Url::parse(x).or(Err(WebDriverError::new(
+                            ErrorStatus::InvalidArgument,
+                            "proxyAutoconfigUrl was not a valid url"))));
+                    },
+                    None => return Err(WebDriverError::new(
+                        ErrorStatus::InvalidArgument,
+                        "proxyAutoconfigUrl was not a string"
+                    ))
+                },
+                "ftpProxy" => try!(self.validate_host_domain("ftpProxy", "ftp", obj, value)),
+                "ftpProxyPort" => try!(self.validate_port("ftpProxyPort", value)),
+                "httpProxy" => try!(self.validate_host_domain("httpProxy", "http", obj, value)),
+                "httpProxyPort" => try!(self.validate_port("httpProxyPort", value)),
+                "socksProxy" => try!(self.validate_host_domain("socksProxy", "ssh", obj, value)),
+                "socksProxyPort" => try!(self.validate_port("socksProxyPort", value)),
+                "socksUsername" => if !value.is_string() {
+                    return Err(WebDriverError::new(ErrorStatus::InvalidArgument,
+                                                   "socksUsername was not a string"))
+                },
+                "socksPassword" => if !value.is_string() {
+                    return Err(WebDriverError::new(ErrorStatus::InvalidArgument,
+                                                   "socksPassword was not a string"))
+                },
+                x => return Err(WebDriverError::new(
+                    ErrorStatus::InvalidArgument,
+                    format!("{} was not a valid proxy configuration capability", x)))
```

Comment:
Should we not ignore unknown extra JSON entries? I think this can have some unfortunate side-effects in the future if the spec changes or the client “sugars” the object with additional keys that are strictly not necessary for us to do anything about.